logic/pulsed/sequence_generator_logic.py

```diff
@@ -1623,6 +1632,60 @@ def _sampling_sequence_sanity_check(self, sequence):
         # Return error code
         return -1 if ensembles_missing else 0
 
+    def _reset_write_benchmark(self):
+        self._write_speed_benchmark['n_benchmarks'] = 0
+        self._write_speed_benchmark['speed_Sas'] = 0
+
+    def _create_pg_device_id_hash(self):
+        """
+        Creates a unique id for the currently used pulse generator.
+        :return: pg_id
+        """
+
+        # indirect creation, as the pulser interface doesn't provide eg. the serial number or model of the device
+        # based on 1. ScalarConstraints in device constraints, 2. config options of device
+        constraint_dict = OrderedDict(
+            (key, val.__dict__) for (key, val) in self.pulsegenerator().get_constraints().__dict__.items() if
+            isinstance(val, ScalarConstraint))
+        id_dict = self.pulsegenerator()._configuration.update(constraint_dict)
```

Comment:
Accessing _configuration might be a bit ugly, but the BaseMixin guarantees that it exists.

Reply:
accessing a private variable from outside the module is always a big no-no and the attribute `_configuration` will not exist in the new qudi core.
Apart from that this line will always cause `id_dict == None`.

Reply:
I agree with the private variable. If we agree that this quick solution is something we want to follow on, we could provide read access to the config or implement unique device / config ids.
Don't see though, why the dict shouldn't work? It does here...

Reply:
Fixed the latter.
I agree on the private variable. If we want to continue on the upload time estimation in a similar way, we could either provide read access or implement unique device/config ids.